tpm/caps.go

```diff
@@ -0,0 +1,73 @@
+package tpm
+
+import (
+	"context"
+	"fmt"
+	"slices"
+
+	"github.com/google/go-tpm/legacy/tpm2"
+	"go.step.sm/crypto/tpm/algorithm"
+)
+
+type Capabilities struct {
+	Algorithms []algorithm.Algorithm
+}
+
+func (c *Capabilities) SupportsAlgorithms(algs ...tpm2.Algorithm) bool {
+	if len(algs) == 0 {
+		return false
+	}
+
+	for _, alg := range algs {
+		if !slices.Contains(c.Algorithms, algorithm.Algorithm(alg)) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Capabilities returns the capabilities of the TPM
+//
+// # Current supports enumerating the supported TPM algorithms
+//
+// # Experimental
+//
+// Notice: This API is EXPERIMENTAL and may be changed or removed in a later
+// release.
+func (t *TPM) GetCapabilities(ctx context.Context) (caps *Capabilities, err error) {
+	caps = &Capabilities{}
+
```

Comment:
You can make it store the capabilities for a future call, and return early if already obtained from the TPM. For a given TPM it doesn't change over time. It'll save a few tx/rx operations, especially if it sits in the path of creating keys. `tpm.Info` does something similar.

Reply:
665a65a